File: 3.x.x/crates-utopia/queue/src/connection/locking.rs
use std::sync::Arc;

use serde_json::Value;

use super::{Connection, StoredValue};
use crate::error::QueueError;
use crate::lock::{Lock, MutexLock};

/// Wraps any [`Connection`] and serializes every command behind a single lock.
///
/// PHP `Utopia\Queue\Connection\Locking`. `ACQUIRE_TIMEOUT = -1` (wait forever).
pub struct Locking<L: Lock = MutexLock> {
    connection: Arc<dyn Connection>,
    lock: L,
}

impl<L: Lock> std::fmt::Debug for Locking<L> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Locking").finish_non_exhaustive()
    }
}

impl Locking<MutexLock> {
    pub fn new(connection: impl Connection + 'static) -> Self {
        Self::with_lock(connection, MutexLock::new())
    }
}

impl<L: Lock> Locking<L> {
    pub const ACQUIRE_TIMEOUT: f64 = -1.0;

    pub fn with_lock(connection: impl Connection + 'static, lock: L) -> Self {
        Self {
            connection: Arc::new(connection),
            lock,
        }
    }

    pub fn from_parts(connection: Arc<dyn Connection>, lock: L) -> Self {
        Self { connection, lock }
    }

    fn synchronize<R>(&self, command: impl FnOnce() -> R) -> R {
        self.lock
            .with_lock(command, Self::ACQUIRE_TIMEOUT)
            .expect("ACQUIRE_TIMEOUT=-1 waits forever")
    }
}

impl<L: Lock> Connection for Locking<L> {
    fn right_push_array(&self, queue: &str, payload: &Value) -> Result<bool, QueueError> {
        self.synchronize(|| self.connection.right_push_array(queue, payload))
    }

    fn right_pop_array(&self, queue: &str, timeout: i64) -> Result<Option<Value>, QueueError> {
        self.synchronize(|| self.connection.right_pop_array(queue, timeout))
    }

    fn right_pop_left_push_array(
        &self,
        queue: &str,
        destination: &str,
        timeout: i64,
    ) -> Result<Option<Value>, QueueError> {
        self.synchronize(|| {
            self.connection
                .right_pop_left_push_array(queue, destination, timeout)
        })
    }

    fn left_push_array(&self, queue: &str, payload: &Value) -> Result<bool, QueueError> {
        self.synchronize(|| self.connection.left_push_array(queue, payload))
    }

    fn left_pop_array(&self, queue: &str, timeout: i64) -> Result<Option<Value>, QueueError> {
        self.synchronize(|| self.connection.left_pop_array(queue, timeout))
    }

    fn right_push(&self, queue: &str, payload: &str) -> Result<bool, QueueError> {
        self.synchronize(|| self.connection.right_push(queue, payload))
    }

    fn right_pop(&self, queue: &str, timeout: i64) -> Result<Option<String>, QueueError> {
        self.synchronize(|| self.connection.right_pop(queue, timeout))
    }

    fn right_pop_left_push(
        &self,
        queue: &str,
        destination: &str,
        timeout: i64,
    ) -> Result<Option<String>, QueueError> {
        self.synchronize(|| {
            self.connection
                .right_pop_left_push(queue, destination, timeout)
        })
    }

    fn left_push(&self, queue: &str, payload: &str) -> Result<bool, QueueError> {
        self.synchronize(|| self.connection.left_push(queue, payload))
    }

    fn left_pop(&self, queue: &str, timeout: i64) -> Result<Option<String>, QueueError> {
        self.synchronize(|| self.connection.left_pop(queue, timeout))
    }

    fn list_remove(&self, queue: &str, key: &str) -> Result<bool, QueueError> {
        self.synchronize(|| self.connection.list_remove(queue, key))
    }

    fn list_size(&self, key: &str) -> Result<i64, QueueError> {
        self.synchronize(|| self.connection.list_size(key))
    }

    fn list_range(&self, key: &str, total: i64, offset: i64) -> Result<Vec<Value>, QueueError> {
        self.synchronize(|| self.connection.list_range(key, total, offset))
    }

    fn remove(&self, key: &str) -> Result<bool, QueueError> {
        self.synchronize(|| self.connection.remove(key))
    }

    fn set(&self, key: &str, value: &str, ttl: i64) -> Result<bool, QueueError> {
        self.synchronize(|| self.connection.set(key, value, ttl))
    }

    fn get(&self, key: &str) -> Result<Option<StoredValue>, QueueError> {
        self.synchronize(|| self.connection.get(key))
    }

    fn set_array(&self, key: &str, value: &Value, ttl: i64) -> Result<bool, QueueError> {
        self.synchronize(|| self.connection.set_array(key, value, ttl))
    }

    fn increment(&self, key: &str) -> Result<i64, QueueError> {
        self.synchronize(|| self.connection.increment(key))
    }

    fn decrement(&self, key: &str) -> Result<i64, QueueError> {
        self.synchronize(|| self.connection.decrement(key))
    }

    fn ping(&self) -> Result<bool, QueueError> {
        self.synchronize(|| self.connection.ping())
    }

    fn close(&self) {
        self.synchronize(|| self.connection.close());
    }
}

File: packages/queue/src/Queue/Connection/Locking.php
<?php

namespace Utopia\Queue\Connection;

use Utopia\Lock\Lock;
use Utopia\Lock\Mutex;
use Utopia\Queue\Connection;

/**
 * Wraps any {@see Connection} and serializes every command behind a single
 * lock, so that concurrent coroutines sharing one connection cannot interleave
 * their requests and responses on the same socket.
 *
 * Outside of a coroutine there is no preemption, so the lock degrades to a
 * plain in-process flag (see {@see Mutex}).
 */
class Locking implements Connection
{
    /**
     * Wait forever when acquiring the lock; a command should never be dropped
     * just because the connection is momentarily busy.
     */
    private const float ACQUIRE_TIMEOUT = -1;

    public function __construct(
        protected readonly Connection $connection,
        protected readonly Lock $lock = new Mutex(),
    ) {}

    /**
     * Run a command while holding the lock, ensuring only one runs at a time.
     *
     * @template T
     * @param callable(): T $command
     * @return T
     */
    protected function synchronize(callable $command): mixed
    {
        return $this->lock->withLock($command, self::ACQUIRE_TIMEOUT);
    }

    public function execute(string $script, array $keys, array $args): mixed
    {
        return $this->synchronize(fn(): mixed => $this->connection->execute($script, $keys, $args));
    }

    public function rightPushArray(string $queue, array $payload): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->rightPushArray($queue, $payload));
    }

    public function rightPopArray(string $queue, int $timeout): array|false
    {
        return $this->synchronize(fn(): array|false => $this->connection->rightPopArray($queue, $timeout));
    }

    public function rightPopLeftPushArray(string $queue, string $destination, int $timeout): array|false
    {
        return $this->synchronize(fn(): array|false => $this->connection->rightPopLeftPushArray($queue, $destination, $timeout));
    }

    public function leftPushArray(string $queue, array $payload): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->leftPushArray($queue, $payload));
    }

    public function leftPushMany(string $queue, array $payloads): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->leftPushMany($queue, $payloads));
    }

    public function rightPushMany(string $queue, array $payloads): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->rightPushMany($queue, $payloads));
    }

    public function leftPopArray(string $queue, int $timeout): array|false
    {
        return $this->synchronize(fn(): array|false => $this->connection->leftPopArray($queue, $timeout));
    }

    public function rightPush(string $queue, string $payload): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->rightPush($queue, $payload));
    }

    public function rightPop(string $queue, int $timeout): string|false
    {
        return $this->synchronize(fn(): string|false => $this->connection->rightPop($queue, $timeout));
    }

    public function rightPopMany(string $queue, int $count, int $timeout): array
    {
        return $this->synchronize(fn(): array => $this->connection->rightPopMany($queue, $count, $timeout));
    }

    public function rightPopLeftPush(string $queue, string $destination, int $timeout): string|false
    {
        return $this->synchronize(fn(): string|false => $this->connection->rightPopLeftPush($queue, $destination, $timeout));
    }

    public function leftPush(string $queue, string $payload): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->leftPush($queue, $payload));
    }

    public function leftPop(string $queue, int $timeout): string|false
    {
        return $this->synchronize(fn(): string|false => $this->connection->leftPop($queue, $timeout));
    }

    public function listRemove(string $queue, string $key): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->listRemove($queue, $key));
    }

    public function listSize(string $key): int
    {
        return $this->synchronize(fn(): int => $this->connection->listSize($key));
    }

    public function listRange(string $key, int $total, int $offset): array
    {
        return $this->synchronize(fn(): array => $this->connection->listRange($key, $total, $offset));
    }

    public function remove(string $key): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->remove($key));
    }

    public function set(string $key, string $value, int $ttl = 0): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->set($key, $value, $ttl));
    }

    public function setNotExists(string $key, string $value, int $ttl = 0): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->setNotExists($key, $value, $ttl));
    }

    public function get(string $key): array|string|null
    {
        return $this->synchronize(fn(): string|array|null => $this->connection->get($key));
    }

    public function setArray(string $key, array $value, int $ttl = 0): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->setArray($key, $value, $ttl));
    }

    public function increment(string $key): int
    {
        return $this->synchronize(fn(): int => $this->connection->increment($key));
    }

    public function incrementBy(string $key, int $by): int
    {
        return $this->synchronize(fn(): int => $this->connection->incrementBy($key, $by));
    }

    public function decrement(string $key): int
    {
        return $this->synchronize(fn(): int => $this->connection->decrement($key));
    }

    public function ping(): bool
    {
        return $this->synchronize(fn(): bool => $this->connection->ping());
    }

    public function close(): void
    {
        $this->synchronize(fn() => $this->connection->close());
    }
}
